src/basic/fd-util.h

```diff
@@ -106,3 +106,20 @@ static inline int make_null_stdio(void) {
         })
 
 int fd_reopen(int fd, int flags);
+
+struct _funlockfile_struct_ {
+        FILE *f;
+        bool quit;
+};
+
+static inline void _release_funlockfile_(struct _funlockfile_struct_ *s) {
+        funlockfile(s->f);
+}
+
+#define WITH_FLOCKFILE(f) \
+        for (_cleanup_(_release_funlockfile_) struct _funlockfile_struct_ _flocked_ = ({ \
```

Comment:
The problem with this construct is that gcc does not understand that the code will run exactly once. As a result, the logic to determine which variables are initialized and such is broken. This happens with the related code for umask a lot, with the result that at `-O2` or `-O3` level there are spurious warnings from seemingly unrelated areas. The initial confusion about the code always running seems to make gcc think that we return success without setting output variables and this cause the issue to "propagate". So… I'm not convinced this a change for the better, even though I agree that the code is conceptually nicer.

Reply:
[This](https://www.chiark.greenend.org.uk/~sgtatham/mp/#S2) looks promising... Doesn't need a boolean to run the loop only once.

Something like this should work:

```c
#define WITH_FLOCKFILE(f) \
    if (0) \
        finished ## __LINE__: ; \
    else \
        for (_cleanup_(_release_funlockfile_) FILE *tmp = ({flockfile(f); f;}) ;;) \
            if (1) \
                goto body ## __LINE__; \
            else \
                while (1) \
                    if (1) \
                        goto finished ## __LINE__; \
                    else \
                        body ## __LINE__:
```

And `_release_funlockfile_` now takes a `FILE **`.

Do you think this might address the spurious warnings from compiling with `-O2` or `-O3`? The code is more complex, but it's all static, so a good compiler should be able to figure this out...

Reply:
Dunno. It would be interesting to see what assembly gcc generates from that. If it is able to simplify all this away, then this would be worth considering. But if it keeps the jumps, I'd rather avoid this.

Reply:
neat hack with the "goto"! if that works, that'd be excellent. It's a fun C excercise, this thing...

Reply:
I'll do some testing on it... Try to build with -O0, -O2 and -O3 and look at the generated code.

I'm actually leaning towards importing [`mp.h`](https://www.chiark.greenend.org.uk/~sgtatham/mp/mp.h) to systemd tree and using the friendly macros in there.

Will keep you posted.

Reply:
See #8946 to possibly supercede this PR.